In [1]:
import pandas as pd
import plotly.express as px
import matplotlib as plt
import datetime

from sqlalchemy import create_engine


In [2]:
col_names = ['TAC Level', 'IR Voltage', 'Temperature', 'Time', 'Date']



In [4]:
df = pd.read_excel(r"C:\Users\ALIYA\OneDrive\Documents\Coding_Temple\Week 5\Day4\data\raw_tac\HV0618 CAM Results.xlsx")

df = df.drop(columns= ['Unnamed: 1','Unnamed: 2', 'Unnamed: 4', 'Unnamed: 5', 'Unnamed: 6',
                   'Unnamed: 7', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 12', 'Unnamed: 13',
                     'Unnamed: 14'])
df.columns = df.iloc[0]
df = df[1:]
df = df.sort_values('TAC Level', ascending=False).reset_index(drop=True)
type(df)

pandas.core.frame.DataFrame

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54 entries, 0 to 53
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   TAC Level    54 non-null     object
 1   IR Voltage   54 non-null     object
 2   Temperature  54 non-null     object
 3   Time         54 non-null     object
 4   Date         54 non-null     object
dtypes: object(5)
memory usage: 2.2+ KB


In [6]:
df.columns

Index(['TAC Level', 'IR Voltage', 'Temperature', 'Time', 'Date'], dtype='object', name=0)

In [8]:
df['Time'] = pd.to_datetime(df['Time'])
df['Date'] = pd.to_datetime(df['Date'])


In [9]:
px.histogram(df, x = df['Date'], y = df['TAC Level'], nbins = 50)

Step 6: Upload to a Mysql instance

Easy way:

In [10]:

my_connection = create_engine('mysql+mysqldb://admin:Cybergirl69@localhost/bonfire_124__schema')

In [11]:
my_connection = my_connection.connect()

In [12]:
df.to_sql(con=my_connection, name='alcohol_levels', if_exists= 'replace')

54

In [13]:
pd.read_sql('SELECT * FROM alcohol_levels', con = my_connection)

index TAC Level  IR Voltage  Temperature                Time  \
0       0     0.224       0.887       85.840 2017-05-02 20:25:15   
1       1     0.181       1.037       77.288 2017-05-02 19:54:15   
2       2     0.170       0.887       72.084 2017-05-02 19:20:07   
3       3     0.167       0.887       71.526 2017-05-02 18:49:06   
4       4     0.143       0.887       81.007 2017-05-02 14:55:17   
5       5     0.130       0.887       89.744 2017-05-02 20:56:13   
6       6     0.110       0.887       80.263 2017-05-02 16:33:45   
7       7     0.105       0.887       80.449 2017-05-02 16:01:28   
8       8     0.105       0.887       91.044 2017-05-03 00:32:42   
9       9     0.090       0.887       89.929 2017-05-02 21:58:03   
10     10     0.071       1.004       89.929 2017-05-02 22:28:58   
11     11     0.063       1.029       84.353 2017-05-02 23:30:49   
12     12     0.051       0.887       86.398 2017-05-03 01:03:36   
13     13     0.048       0.887       80.449 2017-05-02 13:53:40   
14     14     0.041       1.064       81.007 2017-05-02 15:26:14   
15     15     0.038       1.013       84.909 2017-05-02 22:59:52   
16     16     0.036       0.887       89.371 2017-05-02 21:27:08   
17     17     0.033       0.887       76.174 2017-05-02 17:44:16   
18     18     0.030       1.034       76.174 2017-05-02 18:15:15   
19     19     0.027       0.996       86.398 2017-05-03 02:05:31   
20     20     0.027       0.887       82.308 2017-05-03 01:34:32   
21     21     0.026       0.860       78.777 2017-05-02 12:46:48   
22     22     0.025       0.887       87.512 2017-05-03 00:01:46   
23     23     0.023       0.887       88.999 2017-05-03 02:36:26   
24     24     0.022       0.887       93.088 2017-05-03 03:07:21   
25     25     0.019       0.887       79.333 2017-05-02 14:24:35   
26     26     0.018       0.887       75.430 2017-05-02 17:08:00   
27     27     0.017       0.887       92.160 2017-05-03 03:38:15   
28     28     0.012       0.887       89.744 2017-05-03 04:08:53   
29     29     0.008       0.887       93.646 2017-05-03 04:39:31   
30     30     0.008       0.916       93.646 2017-05-03 06:11:22   
31     31     0.008       0.887       93.834 2017-05-03 05:10:09   
32     32     0.007       0.916       92.718 2017-05-03 05:40:46   
33     33     0.006       0.916       94.577 2017-05-03 07:43:11   
34     34     0.006       0.916       93.088 2017-05-03 07:12:35   
35     35     0.005       0.902       78.777 2017-05-02 12:41:09   
36     36     0.000       0.916       91.416 2017-05-03 08:13:49   
37     37     0.000       0.916       86.398 2017-05-03 09:15:03   
38     38     0.000       0.916       80.634 2017-05-03 09:45:38   
39     39     0.000       0.916       77.475 2017-05-03 10:16:23   
40     40     0.000       0.916       91.602 2017-05-03 06:42:00   
41     41     0.000       0.916       91.602 2017-05-03 08:44:25   
42     42     0.000       0.789       80.076 2017-05-02 11:27:59   
43     43     0.000       0.846       76.545 2017-05-02 11:58:24   
44     44     0.000       0.887       80.820 2017-05-02 13:22:59   
45     45     0.000       1.194       81.192 2017-05-02 13:17:44   
46     46     0.000       0.870       78.219 2017-05-02 12:35:29   
47     47     0.000       0.926       77.846 2017-05-02 12:29:52   
48     48     0.000       0.816       77.475 2017-05-02 12:24:37   
49     49     0.000       0.832       77.475 2017-05-02 12:19:23   
50     50     0.000       0.872       77.103 2017-05-02 12:14:08   
51     51     0.000       0.873       76.732 2017-05-02 12:08:53   
52     52     0.000       0.874       76.174 2017-05-02 12:03:39   
53     53     0.000       0.916       72.642 2017-05-03 10:46:39   

                  Date  
0  2017-05-02 20:25:15  
1  2017-05-02 19:54:15  
2  2017-05-02 19:20:07  
3  2017-05-02 18:49:06  
4  2017-05-02 14:55:17  
5  2017-05-02 20:56:13  
6  2017-05-02 16:33:45  
7  2017-05-02 16:01:28  
8  2017-05-03 00:32:42  
9  2017-05-02

In [14]:
import mysql.connector

In [15]:
mydb = mysql.connector.connect(username='admin', password='Cybergirl69', host='127.0.0.1')

cur= mydb.cursor()

cur.execute('SHOW DATABASES')
for x in cur:
    print(x)
cur.close()

('aliyah_first_project',)
('bonfire_124__schema',)
('car_dealership',)
('information_schema',)
('mysql',)
('performance_schema',)
('sakila',)
('sys',)
('world',)


True

In [16]:
mydb = mysql.connector.connect(username='admin', password='Cybergirl69', host='127.0.0.1', database = 'bonfire_124__schema')
cur = mydb.cursor()
cur.execute('SHOW TABLES')
for x in cur:
    print(x)
    


('actor',)
('alcohol_levels',)
('automobiles',)
('film',)
('payment',)


In [17]:
cur.execute('SELECT * FROM alcohol_levels')
cur.fetchall()

[(0,
  '0.224',
  0.887,
  85.84,
  datetime.datetime(2017, 5, 2, 20, 25, 15),
  datetime.datetime(2017, 5, 2, 20, 25, 15)),
 (1,
  '0.181',
  1.037,
  77.288,
  datetime.datetime(2017, 5, 2, 19, 54, 15),
  datetime.datetime(2017, 5, 2, 19, 54, 15)),
 (2,
  '0.170',
  0.887,
  72.084,
  datetime.datetime(2017, 5, 2, 19, 20, 7),
  datetime.datetime(2017, 5, 2, 19, 20, 7)),
 (3,
  '0.167',
  0.887,
  71.526,
  datetime.datetime(2017, 5, 2, 18, 49, 6),
  datetime.datetime(2017, 5, 2, 18, 49, 6)),
 (4,
  '0.143',
  0.887,
  81.007,
  datetime.datetime(2017, 5, 2, 14, 55, 17),
  datetime.datetime(2017, 5, 2, 14, 55, 17)),
 (5,
  '0.130',
  0.887,
  89.744,
  datetime.datetime(2017, 5, 2, 20, 56, 13),
  datetime.datetime(2017, 5, 2, 20, 56, 13)),
 (6,
  '0.110',
  0.887,
  80.263,
  datetime.datetime(2017, 5, 2, 16, 33, 45),
  datetime.datetime(2017, 5, 2, 16, 33, 45)),
 (7,
  '0.105',
  0.887,
  80.449,
  datetime.datetime(2017, 5, 2, 16, 1, 28),
  datetime.datetime(2017, 5, 2, 16, 1, 28)),

In [18]:
pd.read_sql('SELECT * FROM alcohol_levels', con=my_connection)

index TAC Level  IR Voltage  Temperature                Time  \
0       0     0.224       0.887       85.840 2017-05-02 20:25:15   
1       1     0.181       1.037       77.288 2017-05-02 19:54:15   
2       2     0.170       0.887       72.084 2017-05-02 19:20:07   
3       3     0.167       0.887       71.526 2017-05-02 18:49:06   
4       4     0.143       0.887       81.007 2017-05-02 14:55:17   
5       5     0.130       0.887       89.744 2017-05-02 20:56:13   
6       6     0.110       0.887       80.263 2017-05-02 16:33:45   
7       7     0.105       0.887       80.449 2017-05-02 16:01:28   
8       8     0.105       0.887       91.044 2017-05-03 00:32:42   
9       9     0.090       0.887       89.929 2017-05-02 21:58:03   
10     10     0.071       1.004       89.929 2017-05-02 22:28:58   
11     11     0.063       1.029       84.353 2017-05-02 23:30:49   
12     12     0.051       0.887       86.398 2017-05-03 01:03:36   
13     13     0.048       0.887       80.449 2017-05-02 13:53:40   
14     14     0.041       1.064       81.007 2017-05-02 15:26:14   
15     15     0.038       1.013       84.909 2017-05-02 22:59:52   
16     16     0.036       0.887       89.371 2017-05-02 21:27:08   
17     17     0.033       0.887       76.174 2017-05-02 17:44:16   
18     18     0.030       1.034       76.174 2017-05-02 18:15:15   
19     19     0.027       0.996       86.398 2017-05-03 02:05:31   
20     20     0.027       0.887       82.308 2017-05-03 01:34:32   
21     21     0.026       0.860       78.777 2017-05-02 12:46:48   
22     22     0.025       0.887       87.512 2017-05-03 00:01:46   
23     23     0.023       0.887       88.999 2017-05-03 02:36:26   
24     24     0.022       0.887       93.088 2017-05-03 03:07:21   
25     25     0.019       0.887       79.333 2017-05-02 14:24:35   
26     26     0.018       0.887       75.430 2017-05-02 17:08:00   
27     27     0.017       0.887       92.160 2017-05-03 03:38:15   
28     28     0.012       0.887       89.744 2017-05-03 04:08:53   
29     29     0.008       0.887       93.646 2017-05-03 04:39:31   
30     30     0.008       0.916       93.646 2017-05-03 06:11:22   
31     31     0.008       0.887       93.834 2017-05-03 05:10:09   
32     32     0.007       0.916       92.718 2017-05-03 05:40:46   
33     33     0.006       0.916       94.577 2017-05-03 07:43:11   
34     34     0.006       0.916       93.088 2017-05-03 07:12:35   
35     35     0.005       0.902       78.777 2017-05-02 12:41:09   
36     36     0.000       0.916       91.416 2017-05-03 08:13:49   
37     37     0.000       0.916       86.398 2017-05-03 09:15:03   
38     38     0.000       0.916       80.634 2017-05-03 09:45:38   
39     39     0.000       0.916       77.475 2017-05-03 10:16:23   
40     40     0.000       0.916       91.602 2017-05-03 06:42:00   
41     41     0.000       0.916       91.602 2017-05-03 08:44:25   
42     42     0.000       0.789       80.076 2017-05-02 11:27:59   
43     43     0.000       0.846       76.545 2017-05-02 11:58:24   
44     44     0.000       0.887       80.820 2017-05-02 13:22:59   
45     45     0.000       1.194       81.192 2017-05-02 13:17:44   
46     46     0.000       0.870       78.219 2017-05-02 12:35:29   
47     47     0.000       0.926       77.846 2017-05-02 12:29:52   
48     48     0.000       0.816       77.475 2017-05-02 12:24:37   
49     49     0.000       0.832       77.475 2017-05-02 12:19:23   
50     50     0.000       0.872       77.103 2017-05-02 12:14:08   
51     51     0.000       0.873       76.732 2017-05-02 12:08:53   
52     52     0.000       0.874       76.174 2017-05-02 12:03:39   
53     53     0.000       0.916       72.642 2017-05-03 10:46:39   

                  Date  
0  2017-05-02 20:25:15  
1  2017-05-02 19:54:15  
2  2017-05-02 19:20:07  
3  2017-05-02 18:49:06  
4  2017-05-02 14:55:17  
5  2017-05-02 20:56:13  
6  2017-05-02 16:33:45  
7  2017-05-02 16:01:28  
8  2017-05-03 00:32:42  
9  2017-05-02

In [19]:
df['TAC Level'] = df['TAC Level'].astype(float)
df['IR Voltage'] = df['IR Voltage'].astype(float)
df['Temperature'] = df['Temperature'].astype(int)

Creating a pipeline

In [21]:
from category_encoders import OrdinalEncoder

from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import make_pipeline
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
import xgboost as xgb

#### Establish baseline:

In [22]:
baseline = [df['TAC Level'].mean()] * len(df)
print(f'Baseline is: {round(baseline[0],2)}')
len(baseline)
len(df)

Baseline is: 0.04


54

### First, Split the data!

In [23]:
# 2-D Feature Matrix, make predictions based off of:
X = df.drop(columns=['TAC Level','Time','Date'])

# 1-D Target Vector, or the values I am trying to predict
y = df['TAC Level']

#### Now, we split into training and testing

In [24]:
X_train, v, y_train, y = train_test_split(X,y, test_size=.2, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(v,y, test_size=.5, random_state=42)

Make Pipeline

In [25]:
model_rf = make_pipeline(
    OrdinalEncoder(),
    RandomForestRegressor(random_state=42)
)

model_rf.fit(X_train, y_train)

Pipeline(steps=[('ordinalencoder', OrdinalEncoder(cols=[], mapping=[])),
                ('randomforestregressor',
                 RandomForestRegressor(random_state=42))])

In [26]:
train_base = [df['TAC Level'].mean()] * len(y_train)

In [27]:
mean_absolute_error(y_train, train_base)

0.041527131782945745

In [28]:
mean_absolute_error(y_train, model_rf.predict(X_train))

0.015120723698781834

In [29]:
mean_absolute_error(y_test, model_rf.predict(X_test))

0.029103230158730172

In [30]:
r2_score(y_train, model_rf.predict(X_train))

0.837792670611072

In [31]:
r2_score(y_test, model_rf.predict(X_test))

0.14287523925903778

### Conclusion
We where able to get our traing data to predict with a .83 accurcy but with our tesing data we had a munch worse outcome. We belive this is mainly beacuse we used a small set of data. If we had a larger set then we would have a more accurte pipeline.